activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -589,25 +589,41 @@ def exec_cache(sql, name, binds)
             @connection.exec_prepared(stmt_key, type_casted_binds)
           end
         rescue ActiveRecord::StatementInvalid => e
-          pgerror = e.cause
+          raise unless is_cached_plan_failure?(e)
 
-          # Get the PG code for the failure.  Annoyingly, the code for
-          # prepared statements whose return value may have changed is
-          # FEATURE_NOT_SUPPORTED.  Check here for more details:
-          # http://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/utils/cache/plancache.c#l573
-          begin
-            code = pgerror.result.result_error_field(PGresult::PG_DIAG_SQLSTATE)
-          rescue
-            raise e
-          end
-          if FEATURE_NOT_SUPPORTED == code
+          # Nothing we can do if we are in a transaction because all commands
+          # will raise InFailedSQLTransaction
+          if in_transaction?
+            raise ActiveRecord::PreparedStatementCacheExpired.new(e.cause.message)
+          else
+            # outside of transactions we can simply flush this query and retry
             @statements.delete sql_key(sql)
```

Comment:
@matthewd I deliberately left this behavior exactly as it was before to make this change as non-invasive as possible whilst still solving the core issue.

I'm not an expert here, the person to ask would be @tenderlove who wrote this original case.

I think that invalidating the entire cache in all cases would be a safe response. The only time this Postgres exception is likely to be raised is on a schema change. These are rare enough that the performance hit of flushing the entire prepared statement cache is a minor issue.